### ISSUE TITLE
[Feature 🔨]: Support errors when filtering (e.g. `TryFilter`)

### ISSUE THREAD

**Issue by danail-branekov** [BooleanCat/go-functional#135]
**Is your feature request related to a problem? Please describe.**

Checking whether an object is applicable to a [`Filter`](https://github.com/BooleanCat/go-functional/blob/7a5305596a3391be8ba2247efea09cea943dd446/it/itx/filter.go#L6) predicate could theorethically yield erorrs. 

For example, in Korifi we wanted to filter service offering that are related to a broker. In order to check whether an offering is related to a broker one need to lookup the broker in the very filter predicate and that may fail.

Currently errors while filtering is not supported (as the filter predicate cannot return errors), tehrefore we ended up with this [ugly code](https://github.com/cloudfoundry/korifi/blob/966a374fb5690d07c4c644d2c27af46d6820db58/api/repositories/service_offering_repository.go#L71-L84) 

**Describe the solution you'd like**

Maybe a `TryFilter` function, similarly to `TryCollect`

**Provide code snippets to show how this new feature might be used.**

`it.TryFilter(predicate_that_returns_bool_and_error).TryCollect()` -> the error returned could accumulate errors from both `Filter` and `Collect`

**Does this incur a breaking change?**

No

**Do you intend to build this feature yourself?**
We could give it a try

@georgethebeatle

**Comment by BooleanCat** [comment]
Thanks @danail-branekov 

So you have an `iter.Seq[V]` that then needs to be raised to that will become an `iter.Seq[V, error]` after filtering?

I suppose there's a use case for both this and and `Map` that raises an `iter.Seq` into an `iter.Seq2`. This could be a good idea. It would end up looking something like (I'm not sure about the name yet):

```go
func FilterError[V any](delegate func(func(V) bool), predicate func(V) (bool, error)) iter.Seq2[V, error] {
	return func(yield func(V) (bool, error)) {
		for value := range delegate {
			if ok, err := predicate(value); err != nil {
				yield(zero, err)
				return
			} else if ok {
				if !yield(value, nil) {
					return
				}
			}
		}
	}
}
```

> [!NOTE]
> It wouldn't be possible to make a chainable version of this for the `itx` package because Go's type system won't let us constrain a type paramter to `error` on a method or introduce a new generic type parameter on a method.

**Comment by BooleanCat** [comment]
We could also have a map iterator with this signature:

```go
func MapExpand[V, W, X any](delegate func(func(V) bool), op func(V) (W, X)) iter.Seq2[W, X]
```

**Comment by BooleanCat** [comment]
Actually that case may already be covered @danail-branekov!

```go
values, err := it.TryCollect(it.Filter2(it.Zip(slices.Values(mySlice), it.Repeat[error](nil)), myFilter))
```

**Comment by BooleanCat** [comment]
@danail-branekov Is this what you had in mind?

https://github.com/BooleanCat/go-functional/pull/136

**Comment by BooleanCat** [comment]
Solved by #136
